cpp/Naive.cpp

```diff
@@ -1,65 +1,413 @@
 #include "Naive.h"
+#include "equillibrium.h"
 
-namespace ahr {
-    namespace stdex = std::experimental;
+#include <utility>
 
-    auto Naive::getMomentSlice(fftw::mdbuffer<3u> &moments, int m) {
-        return stdex::submdspan(moments.to_mdspan(), m, stdex::full_extent, stdex::full_extent);
+namespace {
+    template<class T, class ...Args, size_t ...I>
+    auto forward_to_array_impl(std::index_sequence<I...> indices, Args &&...args) {
+        return std::array<T, indices.size()>{
+                [&args...](size_t index) {
+                    return T{std::forward<Args>(args)...};
+                }(I)...
+        };
     }
 
-    void Naive::init(mdspan<Real, dextents<Dim, 3u>> initialMoments, Dim N, Real initialDT) {
-        this->N = N,
-        M = initialMoments.extent(0),
-        X = initialMoments.extent(1),
-        Y = initialMoments.extent(2);
-        assert(X == Y);
 
-        fftw::mdbuffer<3u> moments{M, X, Y};
-        for (int m = 0; m < M; ++m) {
-            for (int x = 0; x < X; ++x) {
-                for (int y = 0; y < Y; ++y) {
-                    moments(m, x, y) = initialMoments(m, x, y);
-                }
-            }
-        }
+    template<class T, size_t Size, class ...Args>
+    std::array<T, Size> forward_to_array(Args &&...args) {
+        return forward_to_array_impl<T>(std::make_index_sequence<Size>(), std::forward<Args>(args)...);
+    }
+}
+
+namespace ahr {
+    Naive::Naive(std::ostream &out, Dim M, Dim X, Dim Y) : HermiteRunner(out), M(M), X(X), Y(Y) {}
+
+    void Naive::init(Dim N_) {
+        N = N_;
+
+        // Currently assuming X==Y for simplicity, but the code is written generally for the most part.
+        assert(X == Y);
 
-        auto inSubspan = getMomentSlice(moments, 0);
-        auto outSubspan = getMomentSlice(momentsPH, 0);
+        // Plan FFTs both ways
+        fft = fftw::plan<2u>::dft(phi_K, phi_K_New, fftw::FORWARD, fftw::MEASURE);
+        fftInv = fftw::plan<2u>::dft(phi_K_New, phi_K, fftw::BACKWARD, fftw::MEASURE);
 
-        plan = fftw::plan<2u>::dft(inSubspan, outSubspan, fftw::FORWARD, fftw::MEASURE);
+        // Initialize equilibrium values
+        auto [aParEq, phi] = equilibriumOT01(X, Y);
+        fft(phi, phi_K);
+        fft(aParEq, aParEq_K);
 
+        // DEBUG
+        print(aParEq);
+        print(aParEq_K);
+        // Transform moments into phase space
         for (int m = 0; m < M; ++m) {
-            plan(getMomentSlice(moments, m), getMomentSlice(momentsPH, m));
+            if (m == A_PAR) continue;
+            for_each_xy([&](Dim x, Dim y) {
+                moments_K(m, x, y) = 0;
+            });
         }
+
+        // aPar and uekPar
+        for_each_kxky([&](Dim ky, Dim kx) {
+            aParEq_K(kx, ky) /= std::sqrt(KX*KY);
```

Comment:
This is bc not normalizing inside FFTs?

Reply:
Yes. Which reminds me that I need to normalize everywhere else too

Reply:
Actually `bracket` does the normalization